form/src/main/java/org/dhis2/form/ui/FormViewModel.kt

```diff
@@ -190,42 +192,53 @@ class FormViewModel(
         }
     }
 
-    private fun saveLastFocusedItem(rowAction: RowAction) =
-        repository.currentFocusedItem()?.takeIf {
-            it.valueType?.let { valueType -> valueTypeIsTextField(valueType) } ?: false
-        }?.let {
-            val error = checkFieldError(it.valueType, it.value, it.fieldMask)
-            if (error != null) {
-                val action = rowActionFromIntent(
-                    FormIntent.OnSave(it.uid, it.value, it.valueType, it.fieldMask)
-                )
-                repository.updateErrorList(action)
-                StoreResult(
-                    rowAction.id,
-                    ValueStoreResult.VALUE_HAS_NOT_CHANGED
-                )
-            } else {
-                val action = rowActionFromIntent(
-                    FormIntent.OnSave(it.uid, it.value, it.valueType, it.fieldMask)
-                )
-                val result = repository.save(it.uid, it.value, null)
-                repository.updateValueOnList(it.uid, it.value, it.valueType)
-                repository.updateErrorList(action)
-                result
-            }
-        } ?: StoreResult(
-            rowAction.id,
-            ValueStoreResult.VALUE_HAS_NOT_CHANGED
-        )
+    private fun saveLastFocusedItem(rowAction: RowAction) = getLastFocusedTextItem()?.let {
+        val error = checkFieldError(it.valueType, it.value, it.fieldMask)
+        if (error != null) {
+            val action = rowActionFromIntent(
+                FormIntent.OnSave(it.uid, it.value, it.valueType, it.fieldMask)
+            )
+            repository.updateErrorList(action)
+            StoreResult(
+                rowAction.id,
+                ValueStoreResult.VALUE_HAS_NOT_CHANGED
+            )
+        } else {
+            val intent = getSaveIntent(it)
+            val action = rowActionFromIntent(intent)
+            val result = repository.save(it.uid, it.value, action.extraData)
+            repository.updateValueOnList(it.uid, it.value, it.valueType)
+            repository.updateErrorList(action)
+            result
+        }
+    } ?: StoreResult(
+        rowAction.id,
+        ValueStoreResult.VALUE_HAS_NOT_CHANGED
+    )
 
-    fun valueTypeIsTextField(valueType: ValueType): Boolean {
+    fun valueTypeIsTextField(valueType: ValueType, renderType: UiRenderType? = null): Boolean {
         return valueType.isNumeric ||
-            valueType.isText ||
+            valueType.isText && renderType?.isPolygon() != true ||
```

Comment:
shouldn't has also `valueType.isText ||` ?

Reply:
No, for this case we need to return false if the coordinate is a polygon. If we add valueType.isText it will return true for the polygon case